src/language/tsLanguageService.ts

```diff
@@ -0,0 +1,184 @@
+  SignatureInformation,
+  ParameterInformation,
+} from 'vscode-languageserver/node';
+import ts from 'typescript';
+import { TextDocument, Position } from 'vscode-languageserver-textdocument';
+import { readFileSync } from 'fs';
+import { join, basename, dirname } from 'path';
+
+type TypeScriptServiceHost = {
+  getLanguageService(jsDocument: TextDocument): ts.LanguageService;
+  getCompilationSettings(): ts.CompilerOptions;
+  dispose(): void;
+};
+
+// Server folder.
+const serverPath =
+  basename(__dirname) === 'dist'
+    ? dirname(__dirname)
+    : dirname(dirname(__dirname));
+
+// TypeScript library folder.
+const librarPath = join(serverPath, 'node_modules/typescript/lib');
+const contents: { [name: string]: string } = {};
+
+export default class TypeScriptService {
+  _connection: Connection;
+  _host: TypeScriptServiceHost;
+
+  constructor(connection: Connection) {
+    this._host = this._getTypeScriptServiceHost();
+    this._connection = connection;
+  }
+
+  _loadLibrary(name: string) {
+    let content = contents[name];
+    if (typeof content !== 'string' && librarPath) {
+      const libPath = join(librarPath, name); // From source.
+      try {
+        content = readFileSync(libPath).toString();
+      } catch (e) {
+        this._connection.console.error(
+          `Unable to load library ${name} at ${libPath}`
+        );
+        content = '';
+      }
+      contents[name] = content;
+    }
+    return content;
+  }
+
+  _getTypeScriptServiceHost() {
+    const compilerOptions = {
+      allowNonTsExtensions: true,
+      allowJs: true,
+      lib: ['lib.es2020.full.d.ts'], // Should match to lib from tsconfig.json.
+      target: ts.ScriptTarget.Latest,
+      moduleResolution: ts.ModuleResolutionKind.Classic,
+      experimentalDecorators: false,
+    };
+    let currentTextDocument = TextDocument.create('init', 'javascript', 1, '');
+
+    // Create the language service host to allow the LS to communicate with the host.
+    const host: ts.LanguageServiceHost = {
+      getCompilationSettings: () => compilerOptions,
+      getScriptFileNames: () => [currentTextDocument.uri],
+      getScriptKind: () => {
+        return ts.ScriptKind.JS;
+      },
+      getScriptVersion: (fileName: string) => {
+        if (fileName === currentTextDocument.uri) {
+          return String(currentTextDocument.version);
+        }
+        return '1';
+      },
+      getScriptSnapshot: (fileName: string) => {
+        let text = '';
+        if (fileName === currentTextDocument.uri) {
+          text = currentTextDocument.getText();
+        } else {
+          text = this._loadLibrary(fileName);
+        }
+        return {
+          getText: (start, end) => text.substring(start, end),
+          getLength: () => text.length,
+          getChangeRange: () => undefined,
+        };
+      },
+      getCurrentDirectory: () => '',
+      getDefaultLibFileName: () => 'es2020.full', // Should match to lib from tsconfig.json.
```

Comment:
This should probably always match the filename in line 58 – does it make sense to extract this to a constant that’s used in both places?